core/src/main/java/org/infinispan/interceptors/distribution/NonTxDistributionInterceptor.java

```diff
@@ -101,6 +105,59 @@ private Object computeGetReturn(InternalCacheEntry entry, AbstractDataCommand co
    }
 
    @Override
+   public Object visitGetAllCommand(InvocationContext ctx, GetAllCommand command) throws Throwable {
+      Object returnValue = invokeNextInterceptor(ctx, command);
+      if (command.hasFlag(Flag.CACHE_MODE_LOCAL)
+            || command.hasFlag(Flag.SKIP_REMOTE_LOOKUP)
+            || command.hasFlag(Flag.IGNORE_RETURN_VALUES)
+            || !ctx.isOriginLocal()) {
+         return returnValue;
+      }
+      Map<Object, Object> map = returnValue == null ? command.createMap() : (Map<Object, Object>) returnValue;
+      ConsistentHash ch = command.getConsistentHash();
+
+      Set<Object> requestedKeys = new HashSet<>();
+      for (Object key : command.getKeys()) {
+         if (map.get(key) == null) {
+            CacheEntry entry = ctx.lookupEntry(key);
+            if (entry == null || entry.isNull()) {
+               if (!isValueAvailableLocally(ch, key)) {
+                  requestedKeys.add(key);
+               } else {
+                  if (trace) {
+                     log.tracef("Not doing a remote get for key %s since entry is "
+                           + "mapped to current node (%s) or is in L1. Owners are %s",
+                           toStr(key), rpcManager.getAddress(), ch.locateOwners(key));
+                  }
+                  InternalCacheEntry localEntry = localGetCacheEntry(ctx, key, false, command);
+                  // TODO: shouldn't we copy the entry as GetManyCommand.perform does?
+                  map.put(key, command.isReturnEntries() ? localEntry : localEntry != null ? localEntry.getValue()
+                        : null);
+               }
+            }
+         }
+      }
+      if (!requestedKeys.isEmpty()) {
+         if (trace) {
+            log.tracef("Fetching entries for keys %s from remote nodes", requestedKeys);
+         }
+         Map<Object, InternalCacheEntry> previouslyRetrieved = command.getRemotelyFetched();
+         Map<Object, InternalCacheEntry> justRetrieved = retrieveFromRemoteSources(
+               requestedKeys, ctx, command.getFlags());
+         if (previouslyRetrieved != null) {
+            previouslyRetrieved.putAll(justRetrieved);
+         } else {
+            command.setRemotelyFetched(justRetrieved);
+         }
+         for (Entry<Object, InternalCacheEntry> entry : justRetrieved.entrySet()) {
```

Comment:
It seems the only loss I can see in performance is that if a remote node has a newer topology it won't even try to retrieve values as it would immediately throw the OutdatedTopologyException, which depending on the case could be good or bad.